From 11416ec0bf05ba3c35536500a9e29a62064fb314 Mon Sep 17 00:00:00 2001
From: Lionel Xu <Lionel.Xu@freescale.com>
Date: Fri, 9 Dec 2011 14:25:53 +0800
Subject: ENGR00162729 MX6 ESAI: To reset the codec before initing a stream
 playbacking

Sometimes there is no sound after starting a stream playbacking, this problem
can be resolved by resetting external codec at the beginning of startup.

Signed-off-by: Lionel Xu <R63889@freescale.com>
---
 sound/soc/codecs/cs42888.h  |    1 +
 sound/soc/imx/imx-cs42888.c |    9 +++++++--
 2 files changed, 8 insertions(+), 2 deletions(-)

diff --git a/sound/soc/codecs/cs42888.h b/sound/soc/codecs/cs42888.h
index a650c9d..bea8beb 100644
--- a/sound/soc/codecs/cs42888.h
+++ b/sound/soc/codecs/cs42888.h
@@ -19,6 +19,7 @@
  * the .codec_dai field of your machine driver's snd_soc_dai_link structure.
  */
 extern struct snd_soc_dai_driver cs42888_dai;
+#define CS42888_RST 235
 
 /*
  * The ASoC codec device structure for the CS42888.  Assign this structure
diff --git a/sound/soc/imx/imx-cs42888.c b/sound/soc/imx/imx-cs42888.c
index d34be0a..4c9ca98 100644
--- a/sound/soc/imx/imx-cs42888.c
+++ b/sound/soc/imx/imx-cs42888.c
@@ -20,6 +20,7 @@
 #include <linux/delay.h>
 #include <linux/regulator/consumer.h>
 #include <linux/fsl_devices.h>
+#include <linux/gpio.h>
 #include <sound/core.h>
 #include <sound/pcm.h>
 #include <sound/soc.h>
@@ -33,7 +34,6 @@
 #include "imx-esai.h"
 #include "../codecs/cs42888.h"
 
-
 struct imx_priv_state {
 	int hw;
 };
@@ -45,8 +45,13 @@ static int imx_3stack_startup(struct snd_pcm_substream *substream)
 {
 	struct snd_soc_pcm_runtime *rtd = substream->private_data;
 	struct snd_soc_dai *cpu_dai = rtd->cpu_dai;
-	if (!cpu_dai->active)
+
+	if (!cpu_dai->active) {
 		hw_state.hw = 0;
+		gpio_direction_output(CS42888_RST, 0);
+		msleep(100);
+		gpio_direction_output(CS42888_RST, 1);
+	}
 
 	return 0;
 }
-- 
1.7.9.5

